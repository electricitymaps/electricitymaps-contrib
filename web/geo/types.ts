import { Feature, FeatureCollection, MultiPolygon, Polygon } from '@turf/turf';

import { GEO_CONFIG } from './generateWorld';

export type GeoConfig = typeof GEO_CONFIG;

export interface ZoneConfig {
  subZoneNames?: string[];
  bounding_box?: number[][];
  contributors?: string[];
  disclaimer?: string;
  estimation_method?: string;
  generation_only?: boolean;
  parsers?: {
    consumption?: string;
    consumptionForecast?: string;
    generationForecast?: string;
    price?: string;
    production?: string;
    productionPerModeForecast?: string;
    productionPerUnit?: string;
  };
  [key: string]:
    | undefined
    | string[]
    | number[][]
    | string
    | boolean
    | { [key: string]: string };
}

export interface OptimizedZoneConfig {
  contributors?: number[];
  disclaimer?: string;
  estimation_method?: string;
  parsers: boolean;
  subZoneNames?: string[];
  generation_only?: boolean;
  timezone?: string;
}

export interface ZonesConfig {
  [key: string]: ZoneConfig;
}

export interface OptimizedZonesConfig {
  [key: string]: OptimizedZoneConfig;
}

export interface CombinedZonesConfig {
  contributors: string[];
  zones: OptimizedZonesConfig;
}

export interface ExchangeConfig {
  lonlat: [number, number];
  rotation: number;
  [key: string]: [number, number] | number;
  // The following properties are removed from the generated exchange config
  // and are either not used in the app or are provided by the backend instead.
  // capacity?: [number, number];
  // comment?: string;
  // _comment?: string;
  // parsers?: {
  //   exchange?: string;
  //   exchangeForecast?: string;
  // };
}

export interface ExchangesConfig {
  [key: string]: ExchangeConfig;
}

export declare type FeatureProperties = {
  zoneName: string;
  countryKey: string;
  isAggregatedView?: boolean;
  isHighestGranularity?: boolean;
  isCombined?: boolean;
};

export interface WorldFeatureCollection extends FeatureCollection {
  features: Feature<MultiPolygon | Polygon, FeatureProperties>[];
}
