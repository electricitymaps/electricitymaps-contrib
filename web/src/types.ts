import type {
  Feature,
  FeatureCollection,
  Geometry,
  MultiPolygon,
  Polygon,
} from '@turf/turf';
import { LineString, MultiLineString, Point } from 'geojson';
import { EstimationMethods, TimeRange } from 'utils/constants';

export type Maybe<T> = T | null | undefined;

export type ZoneKey = string;

export interface GridState {
  _disclaimer: string;
  createdAt: string;
  datetimes: {
    /** Object representing the grid state at a single point in time */
    [datetimeKey: string]: {
      /** Array of all exchanges */
      e: {
        [key: ZoneKey]: StateExchangeData;
      };
      /** Array of all zones */
      z: {
        [key: ZoneKey]: StateZoneData;
      };
    };
  };
}

export interface StateZoneData {
  /** Object representing all production values */
  p: {
    /** Carbon intensity */
    ci?: number | null;
    /** Fossil ratio */
    fr?: number | null;
    /** Renewable ratio */
    rr?: number | null;
  };
  /** Object representing all consumption values */
  c: {
    /** Carbon intensity */
    ci?: number | null;
    /** Fossil ratio */
    fr?: number | null;
    /** Renewable ratio */
    rr?: number | null;
  };
  /** Estimation method */
  em?: EstimationMethods | null;
  /** Estimation percentage */
  ep?: number | null;
  /** Represents if the zone has a outage message or not */
  o?: boolean | null;
}

export interface StateExchangeData {
  /** The carbon intensity of the exchange */
  ci: number;
  /** The net flow of the exchange */
  f: number;
}

export interface ExchangeArrowData {
  co2intensity: number;
  netFlow: number;
  rotation: number;
  lonlat: [number, number];
  key: string;
}

export interface ZoneOverviewForTimePeriod {
  [dateTimeKey: string]: ZoneOverview;
}

export interface ZoneOverview {
  zoneKey: string;
  co2intensity?: number;
  co2intensityProduction?: number;
  stateDatetime: string;
  fossilFuelRatio: number;
  renewableRatio: number;
  estimationMethod?: EstimationMethods;
  estimatedPercentage?: number;
}

export type GenerationType =
  | 'biomass'
  | 'coal'
  | 'gas'
  | 'hydro'
  | 'nuclear'
  | 'oil'
  | 'solar'
  | 'unknown'
  | 'geothermal'
  | 'wind';

export type ElectricityStorageType = 'hydro storage' | 'battery storage';
export type ElectricityStorageKeyType = 'battery' | 'hydro';

export type ElectricityModeType = GenerationType | ElectricityStorageType;

export type Exchange = { [key: string]: number };

export interface ZoneDetail extends ZoneOverview {
  _isFinestGranularity: boolean;
  estimatedPercentage?: number;
  measuredPercentage?: number;
  completenessPercentage?: number;
  // Capacity is only available on hourly details
  capacity?: { [key in ElectricityModeType]: number | null };
  capacitySources?: { [key in ElectricityModeType]: string[] | null };
  dischargeCo2Intensities: { [key in ElectricityStorageKeyType]: number };
  dischargeCo2IntensitySources: { [key in ElectricityStorageKeyType]: string };
  exchange: Exchange;
  exchangeCapacities?: {
    [key: ZoneKey]: number[];
  };
  exchangeCo2Intensities: Exchange;
  fossilFuelRatio: number;
  fossilFuelRatioProduction: number;
  isValid: boolean;
  maxCapacity: number;
  maxDischarge: number;
  maxExport: number;
  maxExportCapacity: number;
  maxImport: number;
  maxImportCapacity: number;
  maxProduction: number;
  maxStorage: number;
  maxStorageCapacity: number;
  price?: {
    value: number;
    currency: string;
    disabledReason?: string;
  };
  production: { [key in GenerationType]: Maybe<number> };
  productionCo2Intensities: { [key in GenerationType]: number };
  productionCo2IntensitySources: { [key in GenerationType]: string };
  renewableRatio: number;
  renewableRatioProduction: number;
  source: string[];
  storage: { [key in ElectricityStorageKeyType]: Maybe<number> };
  totalCo2Consumption: number;
  totalCo2Discharge: number | null;
  totalCo2Export: number | null;
  totalCo2Import: number | null;
  totalCo2NetExchange: number | null;
  totalCo2Production: number;
  totalCo2Storage: number | null;
  totalConsumption: number;
  totalDischarge: number | null;
  totalExport: number | null;
  totalImport: number | null;
  totalProduction: number | null;
  totalStorage: number | null;
}

export interface ZoneDetails {
  hasData: boolean;
  futurePrice: FuturePriceData;
  //TODO Remove from backend, most likely unused
  stateAggregation: 'daily' | 'hourly' | 'monthly' | 'yearly';
  zoneStates: {
    [key: string]: ZoneDetail;
  };
  zoneMessage?: ZoneMessage;
}

export interface ZoneMessage {
  message: string;
  issue?: string;
}

export interface GeometryProperties {
  center: [number, number];
  color: string;
  countryKey: string;
  isAggregatedView: boolean;
  isHighestGranularity: boolean;
  zoneId: string;
  zoneName: string;
}
export interface StateGeometryProperties {
  center?: [number, number];
  stateName?: string;
  stateId?: string;
}

export interface MapGeometries extends FeatureCollection<Geometry> {
  features: Array<MapGeometry>;
}

export interface StatesGeometries extends FeatureCollection<Geometry> {
  features: Array<StatesGeometry>;
}
export interface MapGeometry extends Feature<Polygon | MultiPolygon> {
  geometry: MultiPolygon | Polygon;
  Id?: number;
  properties: GeometryProperties;
}

export interface StatesGeometry extends Feature<LineString | MultiLineString | Point> {
  geometry: LineString | MultiLineString | Point;
  Id?: number;
  properties: StateGeometryProperties;
}

export interface MapTheme {
  co2Scale: {
    steps: number[];
    colors: string[];
  };
  oceanColor: string;
  strokeWidth: number;
  strokeColor: string;
  stateBorderColor: string;
  clickableFill: string;
  nonClickableFill: string;
}

export interface FuturePriceData {
  entryCount: number;
  priceData: {
    [key: string]: number;
  };
  currency: string;
  source: string;
  zoneKey: ZoneKey;
}

// Type for the URL parameters that determine app state
export type RouteParameters = {
  resolution?: string;
  zoneId?: string;
  urlTimeRange?: TimeRange;
  urlDatetime?: string;
};
