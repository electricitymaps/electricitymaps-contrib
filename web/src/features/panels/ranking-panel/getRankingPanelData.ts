import { getCountryName, getFullZoneName, getZoneName } from 'translation/translation';
import type { GridState, StateZoneData, ZoneKey } from 'types';
import { SpatialAggregate } from 'utils/constants';
import { getCarbonIntensity } from 'utils/helpers';

import { getHasSubZones, isGenerationOnlyZone } from '../zone/util';
import { ZoneRowType } from './ZoneList';

function filterZonesBySpatialAggregation(
  zoneKey: ZoneKey,
  spatialAggregation: string
): boolean {
  const hasSubZones = getHasSubZones(zoneKey);
  const isSubZone = zoneKey ? zoneKey.includes('-') : true;
  const isCountryView = spatialAggregation === SpatialAggregate.COUNTRY;
  if (isCountryView && isSubZone) {
    return false;
  }
  if (!isCountryView && hasSubZones) {
    return false;
  }
  return true;
}

export const getRankedState = (
  data: GridState | undefined,
  getCo2colorScale: (co2intensity: number) => string,
  sortOrder: 'asc' | 'desc',
  datetimeIndex: string,
  isConsumption: boolean,
  spatialAggregation: string
): ZoneRowType[] => {
  if (!data) {
    return [];
  }
  const gridState = data.datetimes[datetimeIndex];

  if (!gridState || !gridState.z) {
    return [];
  }

  const zoneState = Object.entries(gridState.z) as [keyof GridState, StateZoneData][];

  const orderedZones: ZoneRowType[] = [];

  zoneState.sort((a, b) => {
    // Sort by carbon intensity
    const aCarbonIntensity = getCarbonIntensity(a[1], isConsumption) || 0;
    const bCarbonIntensity = getCarbonIntensity(b[1], isConsumption) || 0;
    return sortOrder === 'asc'
      ? aCarbonIntensity - bCarbonIntensity
      : bCarbonIntensity - aCarbonIntensity;
  });

  // Define the ranking of the zones, starting from 1
  let ranking = 1;

  for (const [key, value] of zoneState) {
    const co2intensity = getCarbonIntensity(value, isConsumption);

    // Filter out zones that don't have a carbon intensity value
    // or are generation only zones
    // or don't match the spatial aggregation
    if (
      !co2intensity ||
      isGenerationOnlyZone(key) ||
      !filterZonesBySpatialAggregation(key, spatialAggregation)
    ) {
      continue;
    }

    const fillColor = co2intensity ? getCo2colorScale(co2intensity) : undefined;
    orderedZones.push({
      zoneId: key,
      color: fillColor,
      co2intensity,
      countryName: getCountryName(key),
      zoneName: getZoneName(key),
      fullZoneName: getFullZoneName(key),
      ranking: ranking,
    });
    ranking++; // Increment the ranking for the next zone.
  }

  return orderedZones;
};
