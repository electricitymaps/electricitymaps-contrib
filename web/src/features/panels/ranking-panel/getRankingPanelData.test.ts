import { GridState } from 'types';
import { describe, expect, it } from 'vitest';

import { getRankedState } from './getRankingPanelData';

const data = {
  datetimes: {
    datetimeIndex: {
      z: {
        zone1: { p: { ci: 10 }, c: { ci: 10 } },
        zone2: { p: { ci: 5 }, c: { ci: 5 } },
        zone3: { p: { ci: undefined }, c: { ci: undefined } },
        zone4: { p: { ci: 15 }, c: { ci: 15 } },
        zone5: { p: { ci: Number.NaN }, c: { ci: Number.NaN } },
        zone6: { p: { ci: 25 }, c: { ci: 25 } },
        zone7: { p: { ci: null }, c: { ci: null } },
        zone8: { p: { ci: 25 }, c: { ci: 25 } },
      },
    },
  },
} as unknown as GridState;

describe('getRankedState', () => {
  it('returns an empty array if data is undefined', () => {
    const actual = getRankedState(
      undefined,
      () => 'color',
      'asc',
      'datetimeIndex',
      true,
      'spatialAggregation'
    );
    expect(actual).toEqual([]);
  });

  it('returns an empty array if gridState or gridState.z is undefined', () => {
    const data = {
      datetimes: {
        datetimeIndex: {},
      },
    } as unknown as GridState;
    const actual = getRankedState(
      data,
      () => 'color',
      'asc',
      'datetimeIndex',
      true,
      'spatialAggregation'
    );
    expect(actual).toEqual([]);
  });

  it('sorts the zones by carbon intensity in ascending order', () => {
    const actual = getRankedState(
      data,
      () => 'color',
      'asc',
      'datetimeIndex',
      true,
      'spatialAggregation'
    );
    expect(actual).toMatchSnapshot();
  });

  it('sorts the zones by carbon intensity in descending order', () => {
    const actual = getRankedState(
      data,
      () => 'color',
      'desc',
      'datetimeIndex',
      true,
      'spatialAggregation'
    );
    expect(actual).toMatchSnapshot();
  });
});
