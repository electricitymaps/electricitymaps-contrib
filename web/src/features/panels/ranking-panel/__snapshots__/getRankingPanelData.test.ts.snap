// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getRankedState > sorts the zones by carbon intensity in ascending order 1`] = `
[
  {
    "co2intensity": 5,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 1,
    "zoneId": "zone2",
    "zoneName": "",
  },
  {
    "co2intensity": 10,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 2,
    "zoneId": "zone1",
    "zoneName": "",
  },
  {
    "co2intensity": 15,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 3,
    "zoneId": "zone4",
    "zoneName": "",
  },
  {
    "co2intensity": 25,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 4,
    "zoneId": "zone6",
    "zoneName": "",
  },
  {
    "co2intensity": 25,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 5,
    "zoneId": "zone8",
    "zoneName": "",
  },
]
`;

exports[`getRankedState > sorts the zones by carbon intensity in descending order 1`] = `
[
  {
    "co2intensity": 25,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 1,
    "zoneId": "zone6",
    "zoneName": "",
  },
  {
    "co2intensity": 25,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 2,
    "zoneId": "zone8",
    "zoneName": "",
  },
  {
    "co2intensity": 15,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 3,
    "zoneId": "zone4",
    "zoneName": "",
  },
  {
    "co2intensity": 10,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 4,
    "zoneId": "zone1",
    "zoneName": "",
  },
  {
    "co2intensity": 5,
    "color": "color",
    "countryName": "",
    "fullZoneName": "",
    "ranking": 5,
    "zoneId": "zone2",
    "zoneName": "",
  },
]
`;
